### PR TITLE
VZ-3384.  Move OAM and App operators to VPO

### PR DESCRIPTION
**Pull request by michael-cico**
# Description

Move the VAO and OAM operator installs to the platform operator. 
- Add component level status reporting to CR status field
- Gate overall install status on combination of install job status and individual component status'
- Changes to Component, component registry, HelmComponent, and Helm CLI code to accommodate installs
- Adds a `--dry-run` option to replace the env var
- Adds a `--bom-path` option to allow overriding the default BOM file path for local development and testing

Fixes VZ-3384.

# Checklist 

As the author of this PR, I have:

- [x] Checked that I included or updated copyright and license notices in all files that I altered
- [ ] Added or updated unit tests for any new functions I added
- [ ] Added or updated integration tests if appropriate
- [ ] Added or updated acceptance tests if appropriate

Code reviewer, please confirm this PR:

- [ ] Addressed the requirement and meets the acceptance criteria
- [ ] Does not introduce unrelated or spurious changes
- [ ] Does not introduce any unapproved dependency
- [ ] Makes sense and it easy to understand, and/or difficult areas of code are clearly documented so that they can be understood

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/helm_component.go [hunk 1]
```diff
@@ -63,7 +72,7 @@ type appendOverridesSig func(log *zap.SugaredLogger, releaseName string, namespa
 type resolveNamespaceSig func(ns string) string
 
 // upgradeFuncSig is a function needed for unit test override
-type upgradeFuncSig func(log *zap.SugaredLogger, releaseName string, namespace string, chartDir string, overrideFile string, overrides string, existingValuesFile string) (stdout []byte, stderr []byte, err error)
+type upgradeFuncSig func(log *zap.SugaredLogger, releaseName string, namespace string, chartDir string, wait bool, dryRun bool, overrides string, overrideFiles ...string) (stdout []byte, stderr []byte, err error)
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm_component.go:75]: Captured this as a cleanup item in VZ-3421.

File: platform-operator/controllers/verrazzano/component/helm_component.go [hunk 2]
```diff
@@ -76,19 +85,58 @@ func (h helmComponent) Name() string {
...
+	failed, err := helm.IsReleaseFailed(h.releaseName, resolvedNamespace)
+	if err != nil {
+		return err
+	}
+	if failed {
+		// Chart install failed, reset the chart to start over
+		// NOTE: we'll likely have to put in some more logic akin to what we do for the scripts, see
+		//       reset_chart() in the common.sh script.  Recovering chart state can be a bit difficult, we
+		//       may need to draw on both the 'ls' and 'status' output for that.
+		helm.Uninstall(log, h.releaseName, resolvedNamespace, h.waitForInstall, dryRun)
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm_component.go:112]: Captured these as potential cleanup items in VZ-3421

File: platform-operator/controllers/verrazzano/component/helm_component.go [hunk 3]
```diff
@@ -109,35 +157,9 @@ func (h helmComponent) Upgrade(log *zap.SugaredLogger, client clipkg.Client, ns
...
-	if len(kvs) > 0 {
-		bldr := strings.Builder{}
-		for i, kv := range kvs {
-			if i > 0 {
-				bldr.WriteString(",")
-			}
-			bldr.WriteString(fmt.Sprintf("%s=%s", kv.key, kv.value))
-		}
-		overrides = bldr.String()
+	overridesString, err := h.buildOverridesString(log, client, namespace)
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm_component.go:160]: Captured this as a possible item for VZ-3422.

File: platform-operator/controllers/verrazzano/component/helm_component.go [hunk 4]
```diff
@@ -168,10 +190,64 @@ func (h helmComponent) Upgrade(log *zap.SugaredLogger, client clipkg.Client, ns
...
+				bldr.WriteString(",")
+			}
+			bldr.WriteString(fmt.Sprintf("%s=%s", kv.key, kv.value))
+		}
+		overridesString = bldr.String()
+	}
+	return overridesString, nil
+}
+
+// Not sure what the point of this is?
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm_component.go:239]: ✅

File: platform-operator/controllers/verrazzano/component/helm_component.go [hunk 5]
```diff
@@ -76,19 +85,58 @@ func (h helmComponent) Name() string {
...
+		return err
+	}
+
+	// vz-specific chart overrides file
+	overridesString, err := h.buildOverridesString(log, client, resolvedNamespace, kvs...)
+	if err != nil {
+		return err
+	}
+
+	// Do the upgrade
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/component/helm_component.go:128]: ✅

File: platform-operator/controllers/verrazzano/install.go [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+			r.updateComponentStatus(log, cr, comp.Name(), "Update ready status", installv1alpha1.InstallComplete)
+			continue
+		}
+		r.updateComponentStatus(log, cr, comp.Name(), "Install starting", installv1alpha1.InstallStarted)
+		err := comp.Install(log, r, cr.Namespace, r.DryRun)
+		if err != nil {
+			//log.Errorf("Error installing component %s: %v", comp.Name(), err)
+			return ctrl.Result{Requeue: true, RequeueAfter: 30 * time.Second}, err
+		}
+		r.updateComponentStatus(log, cr, comp.Name(), "Install complete", installv1alpha1.InstallComplete)
```
[the remark this answers is not in the document]
Comment (michael-cico) [reply platform-operator/controllers/verrazzano/install.go:40]: ✅

**Review by pfmackin — DISMISSED** [review]

**Review by pfmackin — APPROVED** [review]

**Pull request merged**